### DISCUSSION TITLE
Is xournalpp reinventing the wheel?

### DISCUSSION THREAD

**Discussion by Alexis-Prel** [xournalpp/xournalpp#3309] [Ideas] [answered]
**TL;DR: xournalpp has awesome killer features, but user-base constantly requests more. Relying on Inkscape (or similar) for selections, copy-pasting, layers, and tools might represent less work than implementing everything from scratch.**

Let me start by saying how much I like xournalpp. It is incredibly handy, and the very active github page proves that this is the result of countless hours of work from the developers. For me the distinctive features of xournalpp that I could not find elsewhere are:

- really nice handling of handwriting
- easy creation of multiple page documents
- insertion of vertical space (genius feature!)

So a big thank you for that.

This post is going to be a little blunt though, as it addresses what I think is the elephant in the room when scrolling through the Issues tickets (577 to this day): about half the tickets are feature requests.
Many of these features have already been implemented in more mature software.
For reference, you can find at the end of this post a list of requests for features that are already available in Inkscape and its plugins. The redundancy does not stop there in my opinion, considering bug tickets that concern tools that are present in both software.

It is conceivable to port xournalpp as a plugin of Inkscape, but xournalpp's handling of multiple-page documents is markedly superior.

Instead, I am wondering if it would be possible to use Inkscape as a backend for supplying basic tools, such as: text, selections, groups, and layers.
That would shift xournalpp's core tasks to what it already does well: the management of multiple page documents and receiving handwritten input.

- Selections: 
    - #3091; 
    - #2394
    - #1838
    - #1422
    - #1292
    - #1220
    - #1169 
    - and more, like naming a selection as a SVG group, selecting elements by color/stroke style, moving selection with arrow keys, etc.
- Copy-pasting:
     - #3229 
     - #2294
     - and more, such as duplicating or cloning objects (clone change upon original's change), copy/paste the style of an element
- Layers handling: 
    -  #3000 
    - #2784 
    - #2750
    - #2494 
    - #1886
    - #1865
    - #1320
- Extra/improved tools: 
    - #3151 
    - #2884 
    - #2802 
    - #2578
    - #2493
    - #2362
    - #1565 
    - #1361 
    - #1231
    - #1092
    - #729
- View:
    - #3158 
    - #2117
    - #1909
- Page layout:
    - #2191
    - #1883 
    - #1096
- Miscellaneous:
    - #2816 
    - #2305 
    - #2301
    - #1896
    - #1323

**Comment by LittleHuba** [comment, marked as answer]
Hi @Alexis-Prel !

Thanks for your honest opinion and all the work you have put into this post. Accumulating a list of issues as yours is quite some work.
And also thank you for addressing the elephant in the room :wink: 

There is definitely an overlap between Inkscape and Xournal++, as both work with graphical editing. You could say the same for every other graphical editing software out there. In some manner even MS Word and Inkscape share quite a lot. That said, it is not as easy to use Inkscape as backend as it seems. The actual handling of the model (rotating the page, adding a stroke, grouping elements) is the least effort. Integrating this feature with the GUI is another story though. What I want to say with this is, that we would not save a lot of effort with this move. We would still need to integrate all the features of Inkscape into our GUI and do all the input handling ourselves.
Further it would require a near rewrite of Xournal++ to integrate Inkscape as such a substantial part of Xournal++. I do currently not see us doing such a task in the near future. There is a ton of other reasons why this would actually mean more work for us than implementing the features in our own code base. If you want to know them, please let me know and I'll compile a more thorough list.

So while your idea is interesting, it does not really solve the underlying issue.

What would help us though, is a community manager. Somebody that takes care of the evergrowing list of tickets and compiles them into a list of problems in decending order of magnitude, so we know what to tackle next. This way, we developers would get more free time to actually implement the requests. Should you be interested, feel free to reach out to us.

**Reply by mohe2015** [reply]
I think the problem is that Inkscape (probably) is not modular enough and ready for embedding. But in a world where applications would be designed like that I think this would actually have been feasible. Still would've not been an easy task.

**Comment by Technius** [comment]
I also thank you for your help here. I just want to comment about the issue situation (@LittleHuba has already discussed the drawbacks of the Inkscape backend idea).

> This post is going to be a little blunt though, as it addresses what I think is the elephant in the room when scrolling through the Issues tickets (577 to this day): about half the tickets are feature requests.

We only have this many open feature requests because we tend to prioritize bug fixes and issues with existing features instead of adding new things: of the 1335 closed issues, 772 concern bugs. We leave feature request issues open (as long as they fit within the scope of the project) because they are "issues" to someone as long as they remain unimplemented, but we don't have to implement every one! To filter out noise, we have labeled feature requests that we have "accepted" with `priority` labels. We try to tackle `priority::high` issues or `difficulty::easy` issues first.

> The redundancy does not stop there in my opinion, considering bug tickets that concern tools that are present in both software.

Yes, I agree that many concerns in Xournal++ and Inkscape overlap. However, since Xournal++ is focused on note-taking, we're only interested in reimplementing features that help facilitate note-taking. This is much less work than trying to adapt the application on top of an older, larger codebase that no one on the core maintenance team knows anything about.

**Comment by Kreijstal** [comment]
if you could make a good note-taking GUI workspace that is half as good as xournalpp let me know. (Because that would be awesome, and I've been looking)
But as far as it seems, it looks like you would have to rewrite Inkscape itself.. yeah, talk about reinventing the wheel.

**Reply by Alexis-Prel** [reply]
Hmm, I think we are talking about different things.

My point was simply that xournalpp's strength is its GUI, but that some useful primitives for new/enhanced features (layers, transformations, selections, exports) are already implemented in Inkscape.

The question was then about which is least effort between using Inkscape as a backend or implementing these features from scratch. LittleHuba gave a convincing argument as to why the second option is better so I don't see any value in arguing further.
